test/test_gradient_2d_pml.py

```diff
@@ -0,0 +1,208 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from copy import deepcopy
+from firedrake import File
+import firedrake as fire
+import spyro
+
+
+class Gradient_mask_for_pml():
+    def __init__(self, Wave_obj=None):
+        if Wave_obj.abc_active is False:
+            pass
+
+        # Gatting necessary data from wave object
+        pad = Wave_obj.abc_pad_length
+        z = Wave_obj.mesh_z
+        x = Wave_obj.mesh_x
+        V = Wave_obj.function_space
+
+        # building firedrake function for mask
+        z_min = -(Wave_obj.length_z)
+        x_min = 0.0
+        x_max = Wave_obj.length_x
+        mask = fire.Function(V)
+        cond = fire.conditional(z < z_min, 1, 0)
+        cond = fire.conditional(x < x_min, 1, cond)
+        cond = fire.conditional(x > x_max, 1, cond)
+        mask.interpolate(cond)
+
+        # saving mask dofs
+        self.mask_dofs = np.where(mask.dat.data[:] > 0.95)
+        print("DEBUG")
+
+    def apply_mask(self, dJ):
+        dJ.dat.data[self.mask_dofs] = 0.0
+        return dJ
+
+
+def check_gradient(Wave_obj_guess, dJ, rec_out_exact, Jm, plot=False):
+    steps = [1e-3]  # step length
+
+    errors = []
+    V_c = Wave_obj_guess.function_space
+    dm = fire.Function(V_c)
+    dm.assign(dJ)
+
+    for step in steps:
+
+        Wave_obj_guess.reset_pressure()
+        c_guess = fire.Constant(2.0) + step*dm
+        Wave_obj_guess.initial_velocity_model = c_guess
+        Wave_obj_guess.forward_solve()
+        misfit_plusdm = rec_out_exact - Wave_obj_guess.receivers_output
+        J_plusdm = spyro.utils.compute_functional(Wave_obj_guess, misfit_plusdm)
+
+        grad_fd = (J_plusdm - Jm) / (step)
+        projnorm = fire.assemble(dJ * dm * fire.dx(scheme=Wave_obj_guess.quadrature_rule))
+
+        error = np.abs(100 * ((grad_fd - projnorm) / projnorm))
+
+        errors.append(error)
+
+    errors = np.array(errors)
+
+    # Checking if error is first order in step
+    theory = [t for t in steps]
+    theory = [errors[0] * th / theory[0] for th in theory]
+    if plot:
+        plt.close()
+        plt.plot(steps, errors, label="Error")
+        plt.plot(steps, theory, "--", label="first order")
+        plt.legend()
+        plt.title(" Adjoint gradient versus finite difference gradient")
+        plt.xlabel("Step")
+        plt.ylabel("Error %")
+        plt.savefig("gradient_error_verification.png")
+        plt.close()
+
+    # Checking if every error is less than 5 percent
+
+    test1 = (abs(errors[-1]) < 5 )
+    print(f"Gradient error less than 5 percent: {test1}")
+    print(f"Error of {errors}")
+
+    # Checking if error follows expected finite difference error convergence
```

Comment:
This comment could have a `TODO` or `FIXME` marking to make it easier to remember that this is a pending implementation.